backend/domain_types/src/types.rs

```
@@ -233,8 +233,188 @@ impl<
+                                                    sub_code: "INVALID_CARD_NUMBER".to_owned(),
+                                                    error_identifier: 400,
+                                                    error_message: "Invalid card number in Apple Pay data".to_owned(),
+                                                    error_object: None,
+                                                })
+                                            )?,
+                                            application_expiration_month: Secret::new(decrypted_data.application_expiration_month),
+                                            application_expiration_year: Secret::new(decrypted_data.application_expiration_year),
+                                            payment_data: payment_method_data::ApplePayCryptogramData {
+                                                online_payment_cryptogram: Secret::new(decrypted_data.payment_data.clone().map(|pd| pd.online_payment_cryptogram).unwrap_or_default()),
+                                                eci_indicator: decrypted_data.payment_data.map(|pd| pd.eci_indicator),
+                                            },
+                                        }
+                                    )
+                                },
+                                None => {
+                                    return Err(report!(ApplicationErrorResponse::BadRequest(ApiError {
+                                        sub_code: "MISSING_APPLE_PAY_DATA".to_owned(),
+                                        error_identifier: 400,
+                                        error_message: "Apple Pay payment data is required".to_owned(),
+                                        error_object: None,
+                                    })));
+                                }
+                            };
+
+                            let payment_method = apple_wallet.payment_method.ok_or_else(|| {
```

Use change context to lift the error don't discard the errors

---

This BadRequest is the root error when apple_wallet.payment_method is missing.